### PR TITLE
Add Dynamic Client support

### LINKED ISSUE #209: dynamic.DynamicClient
Any plan to have dynamic.DynamicClient ?

### PR DESCRIPTION
**Pull request by bobh66**
Ported the DynamicClient from the synchronous kubernetes client and converted to async.  Included docs and examples.

Fixes #209

### CODE REVIEW

File: kubernetes_asyncio/client/rest.py [hunk 1]
```diff
@@ -137,13 +137,15 @@ async def request(self, method, url, query_params=None, headers=None,
 
         # For `POST`, `PUT`, `PATCH`, `OPTIONS`, `DELETE`
         if method in ['POST', 'PUT', 'PATCH', 'OPTIONS', 'DELETE']:
-            if re.search('json', headers['Content-Type'], re.IGNORECASE):
+            if (re.search('json', headers['Content-Type'], re.IGNORECASE) or
+                    headers['Content-Type'] == 'application/apply-patch+yaml'):
                 if headers['Content-Type'] == 'application/json-patch+json':
                     if not isinstance(body, list):
                         headers['Content-Type'] = 'application/strategic-merge-patch+json'
+                request_body = None
```
Comment (tomplus) [inline kubernetes_asyncio/client/rest.py:145]: As is autogenerated let's try to keep a changeset as small as possible. I think is not necessary.

File: kubernetes_asyncio/client/rest.py [hunk 2]
```diff
@@ -137,13 +137,15 @@ async def request(self, method, url, query_params=None, headers=None,
 
         # For `POST`, `PUT`, `PATCH`, `OPTIONS`, `DELETE`
         if method in ['POST', 'PUT', 'PATCH', 'OPTIONS', 'DELETE']:
-            if re.search('json', headers['Content-Type'], re.IGNORECASE):
+            if (re.search('json', headers['Content-Type'], re.IGNORECASE) or
```
Comment (tomplus) [inline kubernetes_asyncio/client/rest.py:140]: This file is generated by the generator. Could you update scripts/rest_client_patch.diff? This patch is applied after generating files.

**Review by tomplus — CHANGES_REQUESTED** [review]
Impressive! Thanks for you contribution.

File: kubernetes_asyncio/dynamic/discovery.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+        if not self._cache.get('version'):
+            try:
+                self._cache['version'] = {
+                    'kubernetes': await self.client.request('get', '/version', serializer=just_json)
+                }
+            except (ValueError, MaxRetryError) as e:
+                if isinstance(e, MaxRetryError) and not isinstance(e.reason, ProtocolError):
+                    raise
+                if not self.client.configuration.host.startswith("https://"):
+                    raise ValueError("Host value %s should start with https:// when talking to HTTPS endpoint" %
```
[the remark this answers is not in the document]
Comment (bobh66) [reply kubernetes_asyncio/dynamic/discovery.py:168]: Good question - the code is copied from the original implementation: https://github.com/kubernetes-client/python/blob/master/kubernetes/base/dynamic/discovery.py#L150 so any error would affect both implementations.  I can change it if you have any suggestions for improvement.

[the remark this answers is not in the document]
Comment (tomplus) [reply kubernetes_asyncio/dynamic/discovery.py:168]: Let's keep it for now.

File: kubernetes_asyncio/client/rest.py (hunk 1 above)
Comment (bobh66) [reply kubernetes_asyncio/client/rest.py]: Done

File: kubernetes_asyncio/test/test_discovery.py
[the remark this answers is not in the document]
Comment (bobh66) [reply kubernetes_asyncio/test/test_discovery.py:1]: Done

File: kubernetes_asyncio/client/rest.py (hunk 2 above)
Comment (bobh66) [reply kubernetes_asyncio/client/rest.py:140]: Done

File: kubernetes_asyncio/dynamic/client.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+                print(e['object'].metadata)
+                # If you want to gracefully stop the stream watcher
+                watcher.stop()
+        """
+        if not watcher:
+            watcher = watch.Watch()
+
+        # Use field selector to query for named instance so the watch parameter is handled properly.
+        if name:
+            field_selector = f"metadata.name={name}"
```
Comment (bobh66) [inline kubernetes_asyncio/dynamic/client.py:215]: I also pulled in the changes from this PR - https://github.com/kubernetes-client/python/pull/2076 - to fix an issue with watching resources by name with the dynamic client.

**Comment by tomplus** [comment]
Looks good to me.

There is a minor issue in e2e tests, could you fix it? Thanks.

**Comment by tomplus** [comment]
Sorry for bothering you again, now flak8 complains...

**Comment by bobh66** [comment]
> Sorry for bothering you again, now flak8 complains...

Sorry, I should have caught those.  Hopefully all is good now.

**Review by tomplus — APPROVED** [review]
Thanks for your contribution.

**Comment by Jean-Daniel** [comment]
Just a nitpick, but I think this commit misses a require change in setup.py.
As it adds a new package, it should be append in the `setup.py` package list.
the `dynamic` package it currently ignored by deployment tools like `pip`.

**Comment by tomplus** [comment]
@Jean-Daniel You are right, thanks! Do you want to submit PR for it?

**Comment by bobh66** [comment]
Should it also be added to kubernetes_asyncio/__init__.py?

https://github.com/tomplus/kubernetes_asyncio/blob/master/kubernetes_asyncio/__init__.py

**Comment by tomplus** [comment]
@bobh66  yes, it'd be beneficial.

**Comment by bobh66** [comment]
Hi @tomplus - any chance we could get a release with these changes?  Happy to help if there is anything I can do.   Thanks!

**Comment by tomplus** [comment]
Sorry for the inconvenience, I'm going to release it over the weekend.

**Comment by tomplus** [comment]
It's been just released, v25.11.0.

**Pull request merged**